src/memory_server.ml

```diff
@@ -278,22 +299,51 @@ let calculate_boot_time_host_free_memory () =
       pages_to_kib (Int64.of_nativeint boot_time_host_free_pages) in
     Int64.mul 1024L boot_time_host_free_kib
 
+let calc_constant_boot_time_host_free_memory constant_count interval =
+  debug "Check boot time host free memory: constant-count=%d check-interval=%f(seconds)" constant_count interval;
+  let rec calc_constant existing_results =
+    let free_memory = calculate_boot_time_host_free_memory () in
+    debug "Calculating boot time host free memory: %Ld" free_memory;
+    if (List.length existing_results) = (constant_count - 1) then
+      if Listext.List.filter_map (fun x -> if x <> free_memory then Some(true) else None) existing_results = [] then begin
+        (* all the results have the same value *)
+        debug "Boot time host free memory constant value: %Ld" free_memory;
+        free_memory
+      end else begin
+        (* Pop the oldest result *)
+        debug "Pop the oldest memory result";
+        let results = (List.tl existing_results) @ [free_memory] in
```

Comment:
If you need a FIFO queue, please implement it as a sub-module and not inline. Below is an efficient implementation but it could be simpler using a single list.

```
module FIFO : sig
  type 'a t = 'a list * 'a list

  val empty : 'a t

  val size : 'a t -> int

  val peek : 'a t -> 'a option

  val push : 'a -> 'a t -> 'a t

  val pop : 'a t -> 'a t
end = struct
  type 'a t = 'a list * 'a list

  let empty = ([], [])

  let size (xs, ys) = List.length xs + List.length ys

  let peek = function
    | x :: xs, _ -> Some x
    | [], [] -> None
    | [], _ -> assert false

  let push y = function
    | [], [] -> ([ y ], [])
    | [], ys -> assert false
    | xs, ys -> (xs, y :: ys)

  let pop = function
    | [], [] -> failwith "fifo is empty"
    | [ x ], ys -> (List.rev ys, [])
    | x :: xs, ys -> (xs, ys)
    | [], ys -> assert false
end
```

Reply:
Is there any existing library in xapi so I can open the module directly?

Reply:
Alternatively, I don't think it is necessary to remember the last _n_ (up to `constant_count`) values. You could just remember the last one and how many times you have seen the same value until now. However, I would add a small tolerance to the definition of "same value".

Something like:
```
let tolerance = 10 in
let rec stable_boot_time_host_free_memory last matches =
  let new_sample = calculate_boot_time_host_free_memory () in
  let matching =
    Option.fold
      ~none:false
      ~some:(function last_sample -> abs (last_sample - new_sample) < tolerance)
      last
  in
  if matching && matches + 1 >= matches_needed then
    new_sample
  else begin
    Thread.delay interval;
    stable_boot_time_host_free_memory (Some new_sample) (matches + 1)
  end
```